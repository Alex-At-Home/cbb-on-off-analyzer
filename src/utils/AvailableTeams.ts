
export type AvailableTeamMeta = {
  team: string,
  year: string,
  gender: string,
  index_template: string
}

export class AvailableTeams {

  private static flatten(
    arr: Array<Array<AvailableTeamMeta>>
  ): Array<AvailableTeamMeta> {
    return arr.reduce(function(acc, v) {
      return acc.concat(v);
    }, [] as Array<AvailableTeamMeta>);
  }

  static getTeams(
    team: string | null, year: string | null, gender: string | null
  ): Array<AvailableTeamMeta> {
    const list = team ?
      (AvailableTeams.byName[team] || ([] as Array<AvailableTeamMeta>)) :
      AvailableTeams.flatten(Object.values(AvailableTeams.byName));

    return list.filter(function(record) {
      return (!team || (team == record.team)) &&
        (!year || (year == record.year)) && (!gender || (gender == record.gender));
    });
  }
  static getTeam(
    team: string, year: string, gender: string
  ): AvailableTeamMeta | null {
      const retVal = AvailableTeams.getTeams(team, year, gender);
      return (retVal.length > 0) ? retVal[0] : null;
  }
  static readonly byName: Record<string, Array<AvailableTeamMeta>> = {
    // Maryland!
    "Maryland": [
      { team: "Maryland", year: "2016/7", gender: "Men", index_template: "maryland" },
      { team: "Maryland", year: "2018/9", gender: "Men", index_template: "misc_conf" },
      { team: "Maryland", year: "2018/9", gender: "Women", index_template: "women_bigten" },
    ],
    // Other BIG
    "Indiana": [
      { team: "Indiana", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Nebraska": [
      { team: "Nebraska", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Wisconsin": [
      { team: "Wisconsin", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Ohio St.": [
      { team: "Ohio St.", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Michigan St.": [
      { team: "Michigan St.", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Iowa": [
      { team: "Iowa", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Michigan": [
      { team: "Michigan", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Purdue": [
      { team: "Purdue", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Illinois": [
      { team: "Illinois", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Penn St.": [
      { team: "Penn St.", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Northwestern": [
      { team: "Northwestern", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Rutgers": [
      { team: "Rutgers", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Minnesota": [
      { team: "Minnesota", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    // Misc
    "Cincinnati": [
      { team: "Cincinnati", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "Kentucky": [
      { team: "Kentucky", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "Louisville": [
      { team: "Louisville", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "Saint Mary's (CA)": [
      { team: "Saint Mary's (CA)", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "Virginia": [
      { team: "Virginia", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "UConn": [
      { team: "UConn", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
  }
}
