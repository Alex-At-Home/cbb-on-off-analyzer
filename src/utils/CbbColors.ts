import { Color } from "chroma-js";
import chroma from "chroma-js";

type CbbColorTuple = [ (val: number) => string, (val: number) => string ];

export class CbbColors {

  /** Utility to pick a color scale based on whether the stat is offensive or defensive */
  static picker(offScale: (val: number) => string, defScale: (val: number) => string) {
    return (val: any, valMeta: string) => {
      const num = val.value as number;
      return ((num == null) || (num == undefined)) ?
        CbbColors.malformedDataColor : //(we'll use this color to indicate malformed data)
        ("off" == valMeta) ? offScale(num) : defScale(num)
        ;
    };
  }
  /** For off/def tables where only the off is used */
  static offOnlyPicker(offScale: (val: number) => string, defScale: (val: number) => string) {
    return (val: any, valMeta: string) => {
      return ("off" == valMeta) ? CbbColors.picker(offScale, offScale)(val, valMeta) : undefined;
    }
  }
  /** For non-off/def tables, single row */
  static varPicker(scaleFn: (val: number) => string, scale: number = 1) {
    return (val: any, valMeta: string) => {
      const num = val.value as number;
      return ((num == null) || (num == undefined)) ?
        CbbColors.malformedDataColor : //(we'll use this color to indicate malformed data)
        scaleFn(num*scale)
        ;
    };
  }

  static readonly malformedDataColor = "#ccCCcc";
  static readonly background = "#ffFFff";

  private static readonly redToGreen = chroma.scale(["red", "#ffFFff", "green"]);
  private static readonly greenToRed = chroma.scale(["green", "#ffFFff", "red"]);
  private static readonly blueToOrange = chroma.scale(["lightblue", "#ffFFff", "orange"]);
  private static readonly whiteToOrange = chroma.scale(["#ffFFff", "orange"]);

  public static readonly alwaysWhite = (val: number) => CbbColors.background;
  public static readonly alwaysBlack = (val: number) => "#000000";

  // Pts/100
  private static readonly pp100Domain = [ 80, 100, 120 ];
  public static readonly off_pp100 = (val: number) => CbbColors.redToGreen.domain(CbbColors.pp100Domain)(val).toString();
  public static readonly def_pp100 = (val: number) => CbbColors.greenToRed.domain(CbbColors.pp100Domain)(val).toString();
  public static readonly pp100: CbbColorTuple = [ CbbColors.off_pp100, CbbColors.def_pp100 ];
  // eFG
  private static readonly eFGDomain = [ 0.4, 0.5, 0.6 ];
  public static readonly off_eFG = (val: number) => CbbColors.redToGreen.domain(CbbColors.eFGDomain)(val).toString();
  public static readonly def_eFG = (val: number) => CbbColors.greenToRed.domain(CbbColors.eFGDomain)(val).toString();
  public static readonly eFG: CbbColorTuple = [ CbbColors.off_eFG, CbbColors.def_eFG ];
  // Assist rate:
  private static readonly astDomain = [ 0.35, 0.50, 0.65 ];
  public static readonly ast_offDef = (val: number) => CbbColors.blueToOrange.domain(CbbColors.astDomain)(val).toString();
  public static readonly ast: CbbColorTuple = [ CbbColors.ast_offDef, CbbColors.ast_offDef ];
  // TO
  private static readonly toDomain = [ 0.1, 0.16, 0.22 ];
  public static readonly off_TO = (val: number) => CbbColors.greenToRed.domain(CbbColors.toDomain)(val).toString();
  public static readonly def_TO = (val: number) => CbbColors.redToGreen.domain(CbbColors.toDomain)(val).toString();
  public static readonly tOver: CbbColorTuple = [ CbbColors.off_TO, CbbColors.def_TO ];
  // OR
  private static readonly orDomain = [ 0.18, 0.27, 0.36 ];
  public static readonly off_OR = (val: number) => CbbColors.redToGreen.domain(CbbColors.orDomain)(val).toString();
  public static readonly def_OR = (val: number) => CbbColors.greenToRed.domain(CbbColors.orDomain)(val).toString();
  public static readonly oReb: CbbColorTuple = [ CbbColors.off_OR, CbbColors.def_OR ];
  // ftr
  private static readonly ftrDomain = [ 0.20, 0.30, 0.40 ];
  public static readonly off_FTR = (val: number) => CbbColors.redToGreen.domain(CbbColors.ftrDomain)(val).toString();
  public static readonly def_FTR = (val: number) => CbbColors.greenToRed.domain(CbbColors.ftrDomain)(val).toString();
  public static readonly ftr: CbbColorTuple = [ CbbColors.off_FTR, CbbColors.def_FTR ];
  // 3P%
  private static readonly fg3PDomain = [ 0.26, 0.33, 0.40 ];
  public static readonly off_3P = (val: number) => CbbColors.redToGreen.domain(CbbColors.fg3PDomain)(val).toString();
  public static readonly def_3P = (val: number) => CbbColors.greenToRed.domain(CbbColors.fg3PDomain)(val).toString();
  public static readonly fg3P: CbbColorTuple = [ CbbColors.off_3P, CbbColors.def_3P ];
  // 2P%
  private static readonly fg2PDomain = [ 0.4, 0.5, 0.6 ];
  public static readonly off_2P = (val: number) => CbbColors.redToGreen.domain(CbbColors.fg2PDomain)(val).toString();
  public static readonly def_2P = (val: number) => CbbColors.greenToRed.domain(CbbColors.fg2PDomain)(val).toString();
  public static readonly fg2P: CbbColorTuple = [ CbbColors.off_2P, CbbColors.def_2P ];
  // 2P% mid
  private static readonly fg2PMidDomain = [ 0.3, 0.4, 0.5 ];
  public static readonly off_2P_mid = (val: number) => CbbColors.redToGreen.domain(CbbColors.fg2PMidDomain)(val).toString();
  public static readonly def_2P_mid = (val: number) => CbbColors.greenToRed.domain(CbbColors.fg2PMidDomain)(val).toString();
  public static readonly fg2P_mid: CbbColorTuple = [ CbbColors.off_2P_mid, CbbColors.def_2P_mid ];
  // 2P% rim
  private static readonly fg2PRimDomain = [ 0.5, 0.6, 0.7 ];
  public static readonly off_2P_rim = (val: number) => CbbColors.redToGreen.domain(CbbColors.fg2PRimDomain)(val).toString();
  public static readonly def_2P_rim = (val: number) => CbbColors.greenToRed.domain(CbbColors.fg2PRimDomain)(val).toString();
  public static readonly fg2P_rim: CbbColorTuple = [ CbbColors.off_2P_rim, CbbColors.def_2P_rim ];
  // Any FG rate
  private static readonly fgrDomain = [ 0.15, 0.33, 0.5 ];
  public static readonly fgr_offDef = (val: number) => CbbColors.blueToOrange.domain(CbbColors.fgrDomain)(val).toString();
  public static readonly fgr: CbbColorTuple = [ CbbColors.fgr_offDef, CbbColors.fgr_offDef ];

  // Around 0, % (red/green):
  private static readonly diff10DomainRedGreen = [ -0.10, 0, 0.10 ];
  public static readonly off_diff10_redGreen = (val: number) => CbbColors.redToGreen.domain(CbbColors.diff10DomainRedGreen)(val).toString();
  public static readonly def_diff10_redGreen = (val: number) => CbbColors.greenToRed.domain(CbbColors.diff10DomainRedGreen)(val).toString();
  public static readonly diff10_redGreen: CbbColorTuple = [ CbbColors.off_diff10_redGreen, CbbColors.def_diff10_redGreen ];
  public static readonly diff10_greenRed: CbbColorTuple = [ CbbColors.def_diff10_redGreen, CbbColors.off_diff10_redGreen ];
  // bigger diff
  private static readonly diff150DomainRedGreen = [ -1.5, 0, 1.5 ];
  public static readonly off_diff150_redGreen = (val: number) => CbbColors.redToGreen.domain(CbbColors.diff150DomainRedGreen)(val).toString();
  public static readonly def_diff150_redGreen = (val: number) => CbbColors.greenToRed.domain(CbbColors.diff150DomainRedGreen)(val).toString();

  // Around 0, indiv pp100 (red/green):
  private static readonly diff10Domainp100RedGreen = [ -10, 0, 10 ];
  public static readonly off_diff10_p100_redGreen = (val: number) => CbbColors.redToGreen.domain(CbbColors.diff10Domainp100RedGreen)(val).toString();
  public static readonly def_diff10_p100_redGreen = (val: number) => CbbColors.greenToRed.domain(CbbColors.diff10Domainp100RedGreen)(val).toString();
  public static readonly diff10_p100_redGreen: CbbColorTuple = [ CbbColors.off_diff10_p100_redGreen, CbbColors.def_diff10_p100_redGreen ];
  // Around 0, team pp100 (red/green):
  private static readonly diff35Domainp100RedGreen = [ -35, 0, 35 ];
  public static readonly off_diff35_p100_redGreen = (val: number) => CbbColors.redToGreen.domain(CbbColors.diff35Domainp100RedGreen)(val).toString();
  public static readonly def_diff35_p100_redGreen = (val: number) => CbbColors.greenToRed.domain(CbbColors.diff35Domainp100RedGreen)(val).toString();
  public static readonly diff35_p100_redGreen: CbbColorTuple = [ CbbColors.off_diff35_p100_redGreen, CbbColors.def_diff35_p100_redGreen ];
  // Around 0 (blue/orange):
  private static readonly diff10DomainBlueOrange = [ -0.10, 0, 0.10 ];
  public static readonly diff10_blueOrange_offDef = (val: number) => CbbColors.blueToOrange.domain(CbbColors.diff10DomainRedGreen)(val).toString();
  public static readonly diff10_blueOrange: CbbColorTuple = [ CbbColors.diff10_blueOrange_offDef, CbbColors.diff10_blueOrange_offDef ];

  // Personal numbers:
  // Assist rate:
  private static readonly p_astDomain = [ 0.0, 0.15, 0.35 ];
  public static readonly p_ast_offDef = (val: number) => CbbColors.blueToOrange.domain(CbbColors.p_astDomain)(val).toString();
  public static readonly p_ast: CbbColorTuple = [ CbbColors.p_ast_offDef, CbbColors.p_ast_offDef ];
  private static readonly p_astBreakdownDomain = [ 0.0, 0.30 ];
  public static readonly p_ast_breakdown = (val: number) => CbbColors.whiteToOrange.domain(CbbColors.p_astBreakdownDomain)(val).toString();
  // Usage rate:
  private static readonly usgDomain = [ 0.10, 0.20, 0.30 ];
  public static readonly usg_offDef = (val: number) => CbbColors.blueToOrange.domain(CbbColors.usgDomain)(val).toString();
  public static readonly usg: CbbColorTuple = [ CbbColors.usg_offDef, CbbColors.usg_offDef ];
  // Personal rebounding
  private static readonly p_orDomain = [ 0.0, 0.06, 0.12 ];
  private static readonly p_drDomain = [ 0.0, 0.15, 0.30 ];
  public static readonly p_off_OR = (val: number) => CbbColors.blueToOrange.domain(CbbColors.p_orDomain)(val).toString();
  public static readonly p_def_OR = (val: number) => CbbColors.redToGreen.domain(CbbColors.p_drDomain)(val).toString();
  public static readonly p_oReb: CbbColorTuple = [ CbbColors.p_off_OR, CbbColors.p_def_OR ];
  public static readonly p_dReb: CbbColorTuple = [ CbbColors.p_def_OR, CbbColors.p_def_OR ];
  // Personal FTR / FC/50
  public static readonly p_ftr: CbbColorTuple = [ CbbColors.off_FTR, CbbColors.alwaysWhite ];
  // Personal TO / STL
  private static readonly stlDomain = [ 0.0, 0.02, 0.05 ];
  public static readonly p_def_TO = (val: number) => CbbColors.blueToOrange.domain(CbbColors.stlDomain)(val).toString();
  public static readonly p_tOver: CbbColorTuple = [ CbbColors.off_TO, CbbColors.p_def_TO ];

  // Personal Rim / Blk
  private static readonly blkDomain = [ -0.10, 0.0, 0.10 ];
  public static readonly p_def_2P_rim = (val: number) => CbbColors.blueToOrange.domain(CbbColors.blkDomain)(val).toString();
  public static readonly p_fg2P_rim: CbbColorTuple = [ CbbColors.off_2P_rim, CbbColors.p_def_2P_rim ];

  // RAPM diags:
  // Correlation matrix
  private static readonly rapmCorrelDomain = [ 0, 0.5, 1.0 ];
  public static readonly rapmCorrel = (val: number) => CbbColors.blueToOrange.domain(CbbColors.rapmCorrelDomain)(val).toString();
  // Collinearity
  private static readonly rapmCollinLineupDomain = [ 0.01, 0.25, 1.0 ];
  private static readonly rapmCollinPlayerDomain = [ -1.0, 0.0, 1.0 ];
  public static readonly rapmCollinLineup = (val: number) => CbbColors.greenToRed.domain(CbbColors.rapmCollinLineupDomain)(val).toString();
  public static readonly rapmCollinPlayer = (val: number) => CbbColors.blueToOrange.domain(CbbColors.rapmCollinPlayerDomain)(val).toString();

  // Positional diags
  private static readonly posDomain = [ -0.50, 0.0, 0.50 ];
  public static readonly posColors = (val: number) => CbbColors.redToGreen.domain(CbbColors.posDomain)(val).toString();
}
