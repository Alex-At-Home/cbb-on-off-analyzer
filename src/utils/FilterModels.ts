import { DateUtils } from "./DateUtils";

/** Typescript limitations - also have to repeat this for ParamPrefixesType */
export class ParamPrefixes {
  static readonly game = "game-"; //(this is on-off / team info)
  static readonly lineup = "lineup-";
  static readonly report = "report-"; //(not used currrently, we re-use lineup)
  static readonly roster = "roster-";
  static readonly player = "player-";
}
export type ParamPrefixesType = "game-" | "lineup-" | "report-" | "roster-" | "player-";

/** The common luck config */
export type LuckParams = {
  base: "baseline" | "season"
};

export type ManualOverride = {
  rowId: string, //(the player key, lineup key, or on/off key)
  statName: string, //(the field that has been changed)
  newVal: number,
  use: boolean
}

/** Common params across all filter types */
export type CommonFilterParams = {
  year?: string,
  team?: string,
  gender?: string,
  minRank?: string,
  maxRank?: string,
  baseQuery?: string,
  filterGarbage?: boolean, //(missing iff "false")
  queryFilters?: string, //(missing iff empty)

  // FOR INTERNAL USE ONLY
  invertBase?: string //special case - the inverse of these two are used in combination with the usual queryFilters/baseQuery
  invertBaseQueryFilters?: string
};

/** Extracts the common params from a superset */
export function getCommonFilterParams(p: CommonFilterParams) {
  return {
    year: p.year,
    team: p.team,
    gender: p.gender,
    minRank: p.minRank,
    maxRank: p.maxRank,
    baseQuery: p.baseQuery,
    filterGarbage: p.filterGarbage,
    queryFilters: p.queryFilters
  };
}

/** Extracts the common leaderboard params from a superset */
export function getCommonLboardFilterParams(p: CommonFilterParams, tier?: string) {
  return {
    tier: tier,
    year: p.year,
    gender: p.gender,
  };
}

/** Combined params for game filtering */
export type GameFilterParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  onQuery?: string,
  offQuery?: string,
  onQueryFilters?: string, //(missing iff empty)
  offQueryFilters?: string, //(missing iff empty)
  autoOffQuery?: boolean
  // Team view
  teamDiffs?: boolean,
  showExtraInfo?: boolean,
  showTeamPlayTypes?: boolean,
  showRoster?: boolean,
  showGrades?: string,
  // Manual override:
  manual?: ManualOverride[],
  showPlayerManual?: boolean,
  showOnBallConfig?: boolean,
  // Global luck adjustments
  luck?: LuckParams, //(missing iff default)
  // Luck adjustments
  onOffLuck?: boolean,
  showOnOffLuckDiags?: boolean,
  showPlayerOnOffLuckDiags?: boolean,
  showPlayerPlayTypes?: boolean,
  // Misc display:
  showInfoSubHeader?: boolean,
  // Filtering of individual view:
  filter?: string,
  sortBy?: string,
  showBase?: boolean,
  showExpanded?: boolean,
  showDiag?: boolean
  possAsPct?: boolean,
  factorMins?: boolean,
  showPosDiag?: boolean,
  calcRapm?: boolean,
  // For leaderboard building:
  getGames?: boolean
};

/** Params for lineup filtering */
export type LineupFilterParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  // These params need to be explicitly merged in buildParamsFromState(true)
  // For sorting in the generated table:
  decorate?: boolean,
  showTotal?: boolean,
  minPoss?: string,
  maxTableSize?: string,
  sortBy?: string,
  // Filtering:
  filter?: string,
  // Luck adjustments
  luck?: LuckParams, //(missing iff default)
  lineupLuck?: boolean,
  showLineupLuckDiags?: boolean,
  // Other features:
  aggByPos?: string,
  showGameInfo?: boolean
};

export type LineupLeaderboardParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  tier?: string, //High, Medium, Low
  conf?: string, //(undefined ==> all conferences)
  minPoss?: string,
  maxTableSize?: string,
  sortBy?: string,
  // Filtering:
  filter?: string,
  lineupFilters?: string,
  // Luck adjustments
  showLineupLuckDiags?: boolean,
  // Query pre-sets
  confOnly?: boolean,
  t100?: boolean
};

export type TeamEditorParams = {
  [P in keyof PlayerLeaderboardParams]?: PlayerLeaderboardParams[P];
} & {
  // Player editor settings:
  showPrevSeasons?: boolean, //(defaults to false)
  offSeason?: boolean, //(defaults to true, else shows current performance - for building all star teams and seeing effect of injury)
  evalMode?: boolean, //(defaults to false, if true will compare the season X with the predictions from the previous offseason)
  alwaysShowBench?: boolean, //(defaults to false)
  superSeniorsBack?: boolean, //(defaults to false)
  // Controls what transfers are shown
  showOnlyTransfers?: boolean, //(defaults to true)
  showOnlyCurrentYear?: boolean, //(defaults to true)
  diffBasis?: string, //JSON representation of the starting point for showing diffs
  // Editor state
  deletedPlayers?: string, //;-separated list
  disabledPlayers?: string,//;-separated list
  addedPlayers?: string, //;-separated list
  editOpen?: string, //;-separated list, <key>|<open-tab>
  overrides?: string; //;-separated list, see TeamEditorUtils.PlayerEditModel
  allEditOpen?: string, //(defaults to undefined, all edit pages open to the tab if true)
};

export type OffseasonLeaderboardParams = {
  year?: string,
  teamView?: string,
  confs?: string,
  evalMode?: boolean,
} & Record<string, string>;

export type PlayerLeaderboardParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  tier?: string,  //All, High, Medium, Low
  conf?: string, //(undefined ==> all conferences)
  minPoss?: string,
  maxTableSize?: string,
  sortBy?: string,
  // Player settings
  posClasses?: string, //(undefined => all positions)
  possAsPct?: boolean,
  factorMins?: boolean,
  useRapm?: boolean,
  // Filtering:
  filter?: string,
  advancedFilter?: string,
  // Misc display:
  showInfoSubHeader?: boolean,
  // Query pre-sets
  confOnly?: boolean,
  t100?: boolean,
  // Transfer info
  transferMode?: string //==true => show only available, vs ==$year show all
};

export type TeamLeaderboardParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  conf?: string, //(undefined ==> all conferences)
  // Lots of settings:
  qualityWeight?: string,
  pinQualityWeight?: string,
  wabWeight?: string,
  pinWabWeight?: string,
  waeWeight?: string,
  pinWaeWeight?: string,
  domWeight?: string,
  pinDomWeight?: string,
  timeWeight?: string
  pinTimeWeight?: string
};


export type TeamReportFilterParams = {
  [P in keyof CommonFilterParams]?: CommonFilterParams[P];
} & {
  // These params need to be explicitly merged in buildParamsFromState(true)
  // For sorting in the generated table:
  sortBy?: string,
  // Filtering:
  filter?: string,
  showOnOff?: boolean,
  showComps?: boolean,
  incRepOnOff?: boolean,
  regressDiffs?: string, //+ve to add that number of 0 samples, -ve to regress to the given sample size
  repOnOffDiagMode?: string //(the number of diagnostic lineups to show, basically 0 or 20:sort order:sort field)
  incRapm?: boolean,
  rapmDiagMode?: string, //"" if disabled, "team" if enabled with nobody expanded, "playerId[;playerId]+" if expanded for players
  rapmPriorMode?: string, //(-1==default==adapative, else the prior weight as 0->1)
  // Luck adjustments
  luck?: LuckParams, //(missing iff default)
  teamLuck?: boolean
  //(there's no luck diags here because we're applying at the lineup level)
};

/** Used to give compile errors if a field is omitted, for fw compat */
export type RequiredTeamReportFilterParams = {
  [P in keyof TeamReportFilterParams]?: TeamReportFilterParams[P];
} & {
  [P in keyof Required<CommonFilterParams>]: CommonFilterParams[P] | undefined;
};

export class ParamDefaults {
  // Game
  static readonly defaultAutoOffQuery = true;
  static readonly defaultPlayerFilter = "";
  static readonly defaultPlayerSortBy = "desc:off_team_poss_pct:baseline";
  static readonly defaultOnOffLuckAdjust = false;
  static readonly defaultOnOffLuckDiagMode = false;
  static readonly defaultTeamShowPlayTypes = false;
  static readonly defaultTeamShowRoster = false;
  static readonly defaultTeamEnabledGrade = "rank:Combo";
  static readonly defaultPlayerShowPlayTypes = false;
  static readonly defaultPlayerDiagMode = false;
  static readonly defaultPlayerShowBase = false;
  static readonly defaultPlayerShowExpanded = false;
  static readonly defaultPlayerPossAsPct = true;
  static readonly defaultPlayerFactorMins = false;
  static readonly defaultPlayerPosDiagMode = false;
  static readonly defaultPlayerCalcRapm = false;
  // Lineup
  static readonly defaultLineupShowTotal = false;
  static readonly defaultLineupDecorate = true;
  static readonly defaultLineupMinPos = "5";
  static readonly defaultLineupMaxTableSize = "50";
  static readonly defaultLineupSortBy = "desc:off_poss";
  static readonly defaultLineupFilter = "";
  static readonly defaultLineupLuckAdjust = false;
  static readonly defaultLineupLuckDiagMode = false;
  static readonly defaultLineupAggByPos = "";
  static readonly defaultLineupShowGameInfo = false;
  // leaderboards
  static readonly defaultTier = "High";
  // Lineup leaderboard
  static readonly defaultLineupLboardMinPos = "20";
  static readonly defaultLineupLboardMaxTableSize = "100";
  static readonly defaultLineupLboardSortBy = "desc:diff_adj_ppp";
  static readonly defaultLineupLboardFilter = "";
  static readonly defaultLineupLboardLuckDiagMode = false;
  // Player leaderboard
  static readonly defaultPlayerLboardMinPos = "20";
  static readonly defaultPlayerLboardMaxTableSize = "100";
  static defaultPlayerLboardSortBy(useRapm: boolean, factorMins: boolean) {
    return useRapm ? (factorMins ? "desc:diff_adj_rapm_prod" : "desc:diff_adj_rapm") : (factorMins ? "desc:off_adj_prod" : "desc:off_adj_rtg");
  }  
  static readonly defaultPlayerLboardFilter = "";
  static readonly defaultPlayerLboardFactorMins = false;
  static readonly defaultPlayerLboardPossAsPct = true;
  static readonly defaultPlayerLboardUseRapm = true;
  // Team leaderboard
  static readonly defaultTeamLboardQualityWeight = "0.5";
  static readonly defaultTeamLboardDomWeight = "0.25";
  static readonly defaultTeamLboardWabWeight = "0.5"; //(don't have decimal places for comparison with "" + 1.0)
  static readonly defaultTeamLboardWaeWeight = "0.25";
  static readonly defaultTeamLboardTimeWeight = "0"; //(don't have decimal places for comparison with "" + 0.0)
  // Report
  static readonly defaultTeamReportSortBy = "desc:off_poss:on";
  static readonly defaultTeamReportFilter = "";
  static readonly defaultShowOnOff = true;
  static readonly defaultShowComps = false;
  static readonly defaultTeamReportIncRepOnOff = false;
  static readonly defaultTeamReportIncRapm = false;
  static readonly defaultTeamReportRegressDiffs = "-2000";
  static readonly defaultTeamReportRepOnOffDiagMode = "0";
  static readonly defaultTeamReportRepOnOffDiagModeIfEnabled = [ "20", "-1", "lineup.off_poss.value" ];
  static readonly defaultTeamReportRapmDiagMode = "";
  static readonly defaultTeamReportRapmPriorMode = "-1";
  static readonly defaultTeamReportLuckAdjust = false;
  // Common
  static readonly defaultTeam = "";
  static readonly defaultYear = DateUtils.mostRecentYearWithData;
  static readonly defaultLeaderboardYear = DateUtils.mostRecentYearWithLboardData; //(takes a while longer to get updated)
  static readonly defaultGender = "Men";
  static readonly defaultMinRank = "0";
  static readonly defaultMaxRank = "400";
  static readonly defaultFilterGarbage = false;
  static readonly defaultQueryFilters = "";
  static readonly defaultLuckConfig: LuckParams = { base: "season" };
};

export type FilterParamsType = GameFilterParams | LineupFilterParams | TeamReportFilterParams;

/** Which API to call and with what object */
export type FilterRequestInfo = {
  context: ParamPrefixesType,
  paramsObj: FilterParamsType,
  includeRoster?: boolean //(if true will fetch the roster, eg one call per page should do this)
};
