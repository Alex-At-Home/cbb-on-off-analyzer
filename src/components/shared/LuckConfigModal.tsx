// React imports:
import React, { useState, useEffect } from "react";

//lodash
import _ from "lodash";

// Bootstrap imports:
import Modal from "react-bootstrap/Modal";
import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Form from "react-bootstrap/Form";
import FormControl from "react-bootstrap/FormControl";

// Utils:
import { LuckParams } from "../../utils/FilterModels";

// External Data Model

type Props = {
  show: boolean;
  onHide: () => void;
  onSave: (luck: LuckParams) => void;
  luck: LuckParams;
  showHelp: boolean;
};
const LuckConfigModal: React.FunctionComponent<Props> = ({
  onSave,
  luck,
  showHelp,
  ...props
}) => {
  useEffect(() => {
    // Update luck when the settings change:
    setGeneralLuckBase(luck.base as "baseline" | "season");
  }, [luck]);

  // State decomposition:
  const [generalLuckBase, setGeneralLuckBase] = useState(
    luck.base as "baseline" | "season"
  );
  return (
    <Modal
      {...props}
      onEntered={() => {
        document.body.style.overflow = "scroll";
      }}
    >
      <Modal.Header closeButton>
        <Modal.Title>Luck Adjustment Config</Modal.Title>&nbsp;
        {showHelp ? (
          <a
            target="_blank"
            href="https://hoop-explorer.blogspot.com/2020/07/luck-adjustment-details.html"
          >
            (?)
          </a>
        ) : null}
      </Modal.Header>
      <Modal.Body>
        <Card className="w-100">
          <Card.Header className="small">General</Card.Header>
          <Card.Body>
            <Form.Check type="radio">
              <Form.Check.Input
                id="generalLuckBaseSeason"
                type="radio"
                checked={generalLuckBase == "season"}
                onChange={() => {
                  setGeneralLuckBase("season");
                  onSave({ base: "season" });
                }}
              />
              <Form.Check.Label htmlFor="generalLuckBaseSeason">
                Regress over season (recommended)
              </Form.Check.Label>
            </Form.Check>
            <Form.Check type="radio">
              <Form.Check.Input
                id="generalLuckBaseBaseline"
                type="radio"
                checked={generalLuckBase == "baseline"}
                onChange={() => {
                  setGeneralLuckBase("baseline");
                  onSave({ base: "baseline" });
                }}
              />
              <Form.Check.Label htmlFor="generalLuckBaseBaseline">
                Regress over baseline query
              </Form.Check.Label>
            </Form.Check>
          </Card.Body>
        </Card>

        <Card className="w-100">
          <Card.Header className="small">3P Luck</Card.Header>
          <Card.Body>
            <Form inline>
              <Form.Label>3P% Defense is &nbsp;&nbsp;</Form.Label>
              <Form.Control as="select" disabled>
                <option>66%</option>
              </Form.Control>
              <Form.Label> &nbsp;&nbsp;Luck</Form.Label>
            </Form>
          </Card.Body>
        </Card>
      </Modal.Body>
      <Modal.Footer>
        <Button variant="primary" onClick={() => props.onHide()}>
          Exit
        </Button>
      </Modal.Footer>
    </Modal>
  );
};
export default LuckConfigModal;
