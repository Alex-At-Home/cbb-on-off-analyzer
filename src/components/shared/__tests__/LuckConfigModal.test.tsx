import React from 'react';
import LuckConfigModal from '../LuckConfigModal';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import { LuckParams } from '../../../utils/FilterModels';

describe("LuckConfigModal", () => {
  test("LuckConfigModal - should create snapshot", () => {
    //(annoyingly this needs to be shallow because Modal doesn't work with full rendering)
    const wrapper = shallow(
      <div>
        <LuckConfigModal
          show={true}
          onHide={() => {}}
          onSave={(l: LuckParams) => {}}
          luck={{base: "baseline"}}
        />
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
