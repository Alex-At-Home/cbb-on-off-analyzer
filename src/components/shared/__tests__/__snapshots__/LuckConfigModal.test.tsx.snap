// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuckConfigModal LuckConfigModal - should create snapshot 1`] = `
<div>
  <LuckConfigModal
    luck={
      Object {
        "base": "baseline",
      }
    }
    onHide={[Function]}
    onSave={[Function]}
    show={true}
    showHelp={true}
  />
</div>
`;
